File: go/ff1.go
package dealcode

// FF1 format-preserving encryption (NIST SP 800-38G, Algorithms 7 and 8),
// implemented directly from the NIST specification and validated against the
// official NIST FF1-AES sample vectors (testvectors/ff1_nist.json).
//
// Numeral strings are represented as []int with each element in [0, radix);
// conversion to and from characters is the caller's concern.

import (
	"crypto/aes"
	"crypto/cipher"
	"encoding/binary"
	"errors"
	"math/big"
)

// ff1 is an FF1-AES context for one key and radix. The embedded cipher.Block
// returned by aes.NewCipher is safe for concurrent use, and every other field
// is read-only after construction, so an ff1 may be shared by any number of
// goroutines.
type ff1 struct {
	block    cipher.Block
	radix    int
	radixBig *big.Int
}

func newFF1(key []byte, radix int) (*ff1, error) {
	if radix < 2 || radix > 1<<16 {
		return nil, errors.New("ff1: radix must be in [2, 2^16]")
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	return &ff1{block: block, radix: radix, radixBig: big.NewInt(int64(radix))}, nil
}

// ff1Params holds every value that depends only on (tweak, message length):
// the split point, byte sizes, the constant prefix of the CBC-MAC input Q
// (P || tweak || zero padding), and the Feistel moduli radix^u and radix^v.
// All fields are read-only after construction.
type ff1Params struct {
	u, v    int      // left/right half lengths: u = n/2, v = n - u
	b       int      // byte length of NUM(B): ceil(ceil(v*log2(radix))/8)
	d       int      // 4*ceil(b/4) + 4
	qPrefix []byte   // P || tweak || zero pad; each round appends [i] || NUM(b bytes)
	modU    *big.Int // radix^u
	modV    *big.Int // radix^v
}

// params computes the per-(tweak, length) constants. It rejects messages
// shorter than 2 numerals or with a domain radix^n < 100 (the FF1 structural
// minimum).
func (f *ff1) params(tweak []byte, n int) (*ff1Params, error) {
	if n < 2 {
		return nil, errors.New("ff1: message must have at least 2 numerals")
	}
	domain := new(big.Int).Exp(f.radixBig, big.NewInt(int64(n)), nil)
	if domain.Cmp(big.NewInt(100)) < 0 {
		return nil, errors.New("ff1: radix^n must be at least 100")
	}
	u := n / 2
	v := n - u
	modU := new(big.Int).Exp(f.radixBig, big.NewInt(int64(u)), nil)
	modV := new(big.Int).Exp(f.radixBig, big.NewInt(int64(v)), nil)
	// b = ceil(ceil(v*log2(radix))/8), computed exactly as the bit length of
	// radix^v - 1 — floating-point log must not be used.
	b := (new(big.Int).Sub(modV, big.NewInt(1)).BitLen() + 7) / 8
	d := 4*((b+3)/4) + 4

	t := len(tweak)
	pad := (((-t - b - 1) % 16) + 16) % 16
	q := make([]byte, 0, 16+t+pad)
	// P = [1] [2] [1] [radix]^3 [10] [u mod 256] [n]^4 [t]^4
	q = append(q, 1, 2, 1, byte(f.radix>>16), byte(f.radix>>8), byte(f.radix), 10, byte(u))
	q = binary.BigEndian.AppendUint32(q, uint32(n))
	q = binary.BigEndian.AppendUint32(q, uint32(t))
	q = append(q, tweak...)
	q = append(q, make([]byte, pad)...)
	return &ff1Params{u: u, v: v, b: b, d: d, qPrefix: q, modU: modU, modV: modV}, nil
}

// roundY computes the round value y = NUM(S[:d]) for round i and the numeral
// value numVal (NUM of the half that feeds the round function). q is a
// caller-owned scratch buffer holding qPrefix in its first prefixLen bytes
// with room for [i] and the b-byte numeral value after it.
func (f *ff1) roundY(q []byte, prefixLen int, round byte, numVal *big.Int, d int, y *big.Int) {
	q[prefixLen] = round
	numVal.FillBytes(q[prefixLen+1:]) // zero-pads on the left
	// R = PRF(Q): CBC-MAC with a zero IV.
	var r, blk [16]byte
	for off := 0; off < len(q); off += 16 {
		for j := 0; j < 16; j++ {
			blk[j] = r[j] ^ q[off+j]
		}
		f.block.Encrypt(r[:], blk[:])
	}
	if d <= 16 {
		// Within dealcode's configuration bounds d <= 16 always holds, so the
		// S expansion below never runs; it is implemented anyway so the FF1
		// core is fully general.
		y.SetBytes(r[:d])
		return
	}
	// S = R || CIPH(R xor [1]^16) || CIPH(R xor [2]^16) || ..., truncated to d.
	s := make([]byte, 0, ((d+15)/16)*16)
	s = append(s, r[:]...)
	var jb, out [16]byte
	for j := uint64(1); len(s) < d; j++ {
		binary.BigEndian.PutUint64(jb[8:], j)
		for k := 0; k < 16; k++ {
			blk[k] = r[k] ^ jb[k]
		}
		f.block.Encrypt(out[:], blk[:])
		s = append(s, out[:]...)
	}
	y.SetBytes(s[:d])
}

// num sets acc to NUM(xs) — the big-endian base-radix value of the numeral
// string — and returns acc. tmp is scratch.
func (f *ff1) num(xs []int, acc, tmp *big.Int) *big.Int {
	acc.SetInt64(0)
	for _, x := range xs {
		acc.Mul(acc, f.radixBig)
		acc.Add(acc, tmp.SetInt64(int64(x)))
	}
	return acc
}

// nstr writes STR(val, radix, len(out)) — val as exactly len(out) big-endian
// base-radix numerals — into out. val and rem are consumed as scratch.
func (f *ff1) nstr(val *big.Int, out []int, rem *big.Int) {
	for i := len(out) - 1; i >= 0; i-- {
		val.QuoRem(val, f.radixBig, rem)
		out[i] = int(rem.Int64())
	}
}

// encrypt is FF1.Encrypt (Algorithm 7) for the precomputed params p.
// x must have exactly p.u+p.v numerals, each in [0, radix).
func (f *ff1) encrypt(p *ff1Params, x []int) []int {
	u, v := p.u, p.v
	// Three rotating numeral buffers; v >= u so capacity v fits either half.
	a := append(make([]int, 0, v), x[:u]...)
	b := append(make([]int, 0, v), x[u:]...)
	t := make([]int, 0, v)

	q := make([]byte, len(p.qPrefix)+1+p.b)
	copy(q, p.qPrefix)

	var numA, numB, y, c, tmp big.Int
	for i := 0; i < 10; i++ {
		f.roundY(q, len(p.qPrefix), byte(i), f.num(b, &numB, &tmp), p.d, &y)
		m, mod := u, p.modU
		if i%2 == 1 {
			m, mod = v, p.modV
		}
		f.num(a, &numA, &tmp)
		c.Add(&numA, &y)
		c.Mod(&c, mod)
		nb := t[:m]
		f.nstr(&c, nb, &tmp)
		a, b, t = b, nb, a[:0]
	}
	out := make([]int, 0, u+v)
	out = append(out, a...)
	return append(out, b...)
}

// decrypt is FF1.Decrypt (Algorithm 8) for the precomputed params p.
// x must have exactly p.u+p.v numerals, each in [0, radix).
func (f *ff1) decrypt(p *ff1Params, x []int) []int {
	u, v := p.u, p.v
	a := append(make([]int, 0, v), x[:u]...)
	b := append(make([]int, 0, v), x[u:]...)
	t := make([]int, 0, v)

	q := make([]byte, len(p.qPrefix)+1+p.b)
	copy(q, p.qPrefix)

	var numA, numB, y, c, tmp big.Int
	for i := 9; i >= 0; i-- {
		f.roundY(q, len(p.qPrefix), byte(i), f.num(a, &numA, &tmp), p.d, &y)
		m, mod := u, p.modU
		if i%2 == 1 {
			m, mod = v, p.modV
		}
		f.num(b, &numB, &tmp)
		c.Sub(&numB, &y)
		c.Mod(&c, mod) // Euclidean: result is always in [0, mod)
		na := t[:m]
		f.nstr(&c, na, &tmp)
		b, a, t = a, na, b[:0]
	}
	out := make([]int, 0, u+v)
	out = append(out, a...)
	return append(out, b...)
}

File: rust/src/ff1.rs
//! FF1 format-preserving encryption (NIST SP 800-38G, Algorithms 7 and 8).
//!
//! Implemented directly from the NIST specification and validated against the
//! official NIST FF1-AES sample vectors (`testvectors/ff1_nist.json`).
//!
//! This module is internal: numeral strings are `&[u8]` slices of values in
//! `[0, radix)` (dealcode caps the radix at 94, so `u8` numerals suffice), and
//! all callers go through configuration validation that guarantees the
//! arithmetic below fits in `u128` (`radix^v < 2^96`, `d ≤ 16`).

use aes::cipher::{BlockEncrypt, KeyInit};
use aes::{Aes128, Aes192, Aes256, Block};

/// AES-ECB block encryption for whichever key size the codec was built with.
///
/// Round keys are expanded once at construction; `encrypt_block` takes `&self`
/// so a single cipher is freely shared across threads.
// The variants intentionally mirror the AES key-size names.
#[allow(clippy::enum_variant_names)]
#[derive(Clone)]
pub(crate) enum AesCipher {
    Aes128(Aes128),
    Aes192(Aes192),
    Aes256(Aes256),
}

impl AesCipher {
    /// Builds a cipher from a 16-, 24-, or 32-byte AES key.
    ///
    /// The key length is validated by the caller (key resolution only ever
    /// produces those three sizes).
    pub(crate) fn new(key: &[u8]) -> Self {
        match key.len() {
            16 => AesCipher::Aes128(Aes128::new(key.into())),
            24 => AesCipher::Aes192(Aes192::new(key.into())),
            32 => AesCipher::Aes256(Aes256::new(key.into())),
            // Key resolution (SPEC §2.1) only produces 16/24/32-byte keys.
            n => unreachable!("AES key must be 16, 24, or 32 bytes, got {n}"),
        }
    }

    fn encrypt_block(&self, block: &mut [u8; 16]) {
        let block = Block::from_mut_slice(block);
        match self {
            AesCipher::Aes128(c) => c.encrypt_block(block),
            AesCipher::Aes192(c) => c.encrypt_block(block),
            AesCipher::Aes256(c) => c.encrypt_block(block),
        }
    }

    /// `PRF(X)` from SP 800-38G: CBC-MAC with a zero IV, built by chaining
    /// ECB block encryptions. `data.len()` is always a multiple of 16 by
    /// construction of `Q`.
    fn prf(&self, data: &[u8]) -> [u8; 16] {
        debug_assert_eq!(data.len() % 16, 0, "PRF input must be whole blocks");
        let mut acc = [0u8; 16];
        for block in data.chunks_exact(16) {
            for (a, b) in acc.iter_mut().zip(block) {
                *a ^= b;
            }
            self.encrypt_block(&mut acc);
        }
        acc
    }
}

/// Precomputed per-(tweak, length) FF1 parameters (step 1-4 of Algorithms
/// 7/8, plus the constant prefix of `Q`).
#[derive(Clone)]
pub(crate) struct Params {
    /// Length of the left half.
    u: usize,
    /// Length of the right half.
    v: usize,
    /// Byte length of `NUM(B)` in `Q`.
    b: usize,
    /// Byte length of the round output `S`.
    d: usize,
    /// `P ‖ tweak ‖ zero padding` — everything in `P ‖ Q` before the round
    /// number byte.
    q_prefix: Vec<u8>,
    /// `radix^u`.
    mod_u: u128,
    /// `radix^v`.
    mod_v: u128,
}

/// Computes FF1 parameters for numeral strings of length `n` under `tweak`.
///
/// Panics if `radix^n < 100`, `n < 2`, or the configuration exceeds the
/// bounds under which `u128` arithmetic is exact (`d ≤ 16`); dealcode codec
/// construction validates all of these before this is ever called.
pub(crate) fn params(radix: u32, tweak: &[u8], n: usize) -> Params {
    let r = u128::from(radix);
    assert!((2..=65536).contains(&radix), "radix out of range [2, 2^16]");
    assert!(n >= 2, "FF1 message too short");
    // radix^n >= 100 (FF1's structural minimum domain); None means overflow,
    // which is certainly >= 100.
    assert!(
        r.checked_pow(n as u32).map_or(true, |x| x >= 100),
        "FF1 domain radix^n must be at least 100"
    );

    let u = n / 2;
    let v = n - u;
    let mod_v = r
        .checked_pow(v as u32)
        .expect("radix^v fits in u128 (guaranteed by radix^max_length <= 2^128)");
    let mod_u = r
        .checked_pow(u as u32)
        .expect("radix^u fits in u128 (u <= v)");

    // b = ceil(ceil(v * log2(radix)) / 8), computed exactly as the bit
    // length of radix^v - 1 (never via floating-point logarithms).
    let b = (128 - (mod_v - 1).leading_zeros() as usize).div_ceil(8);
    let d = 4 * b.div_ceil(4) + 4;
    assert!(d <= 16, "S expansion exceeds 128 bits (unreachable within validated configurations)");

    let t = tweak.len();
    let pad = (16 - (t + b + 1) % 16) % 16;
    let mut q_prefix = Vec::with_capacity(16 + t + pad);
    // P = [1]1 ‖ [2]1 ‖ [1]1 ‖ [radix]3 ‖ [10]1 ‖ [u mod 256]1 ‖ [n]4 ‖ [t]4
    q_prefix.extend_from_slice(&[
        1,
        2,
        1,
        (radix >> 16) as u8,
        (radix >> 8) as u8,
        radix as u8,
        10,
        u as u8,
    ]);
    q_prefix.extend_from_slice(&(n as u32).to_be_bytes());
    q_prefix.extend_from_slice(&(t as u32).to_be_bytes());
    q_prefix.extend_from_slice(tweak);
    q_prefix.resize(q_prefix.len() + pad, 0);

    Params { u, v, b, d, q_prefix, mod_u, mod_v }
}

/// `NUM_radix(xs)`: numeral string to integer, big-endian.
pub(crate) fn num(radix: u128, xs: &[u8]) -> u128 {
    // Callers guarantee radix^len(xs) <= 2^128, so the fold cannot overflow:
    // the accumulator stays < radix^len(xs) - 1 <= 2^128 - 1.
    xs.iter()
        .fold(0u128, |acc, &x| acc.wrapping_mul(radix).wrapping_add(u128::from(x)))
}

/// `STR^len_radix(value)`: integer to a zero-padded numeral string.
fn str_radix(radix: u128, mut value: u128, len: usize) -> Vec<u8> {
    let mut out = vec![0u8; len];
    for slot in out.iter_mut().rev() {
        *slot = (value % radix) as u8;
        value /= radix;
    }
    out
}

/// One round's `y`: `PRF(P ‖ Q)` expanded to `d` bytes and read big-endian.
///
/// The general `S = R ‖ CIPH(R ⊕ [1]16) ‖ CIPH(R ⊕ [2]16) ‖ …` expansion loop
/// is implemented even though `d ≤ 16` never needs it under dealcode's
/// bounds.
fn round_y(cipher: &AesCipher, p: &Params, round: u8, num_val: u128) -> u128 {
    let mut data = Vec::with_capacity(p.q_prefix.len() + 16);
    data.extend_from_slice(&p.q_prefix);
    data.push(round);
    data.extend_from_slice(&num_val.to_be_bytes()[16 - p.b..]);
    let r = cipher.prf(&data);

    let mut s = r.to_vec();
    let mut j: u128 = 1;
    while s.len() < p.d {
        let jb = j.to_be_bytes();
        let mut block = [0u8; 16];
        for (out, (a, b)) in block.iter_mut().zip(r.iter().zip(jb.iter())) {
            *out = a ^ b;
        }
        cipher.encrypt_block(&mut block);
        s.extend_from_slice(&block);
        j += 1;
    }

    s[..p.d]
        .iter()
        .fold(0u128, |acc, &byte| (acc << 8) | u128::from(byte))
}

/// `FF1.Encrypt` (Algorithm 7). `x` is a numeral string; returns one of the
/// same length.
pub(crate) fn encrypt(cipher: &AesCipher, p: &Params, radix: u32, x: &[u8]) -> Vec<u8> {
    let r = u128::from(radix);
    let mut a = x[..p.u].to_vec();
    let mut b_half = x[p.u..].to_vec();
    for i in 0..10u8 {
        let y = round_y(cipher, p, i, num(r, &b_half));
        let (m, modulus) = if i % 2 == 0 { (p.u, p.mod_u) } else { (p.v, p.mod_v) };
        // Reduce y before adding: both operands are < modulus <= 2^96, so the
        // u128 addition cannot overflow.
        let c = (num(r, &a) + y % modulus) % modulus;
        a = std::mem::replace(&mut b_half, str_radix(r, c, m));
    }
    a.extend_from_slice(&b_half);
    a
}

/// `FF1.Decrypt` (Algorithm 8). Inverse of [`encrypt`].
pub(crate) fn decrypt(cipher: &AesCipher, p: &Params, radix: u32, x: &[u8]) -> Vec<u8> {
    let r = u128::from(radix);
    let mut a = x[..p.u].to_vec();
    let mut b_half = x[p.u..].to_vec();
    for i in (0..10u8).rev() {
        let y = round_y(cipher, p, i, num(r, &a));
        let (m, modulus) = if i % 2 == 0 { (p.u, p.mod_u) } else { (p.v, p.mod_v) };
        // Modular subtraction without signed types or overflow: every term
        // is < modulus <= 2^96.
        let c = (num(r, &b_half) + (modulus - y % modulus)) % modulus;
        b_half = std::mem::replace(&mut a, str_radix(r, c, m));
    }
    a.extend_from_slice(&b_half);
    a
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::Deserialize;

    /// The NIST sample alphabet: character index = numeral value.
    const NIST_ALPHABET: &str = "0123456789abcdefghijklmnopqrstuvwxyz";

    #[derive(Deserialize)]
    struct File {
        vectors: Vec<Vector>,
    }

    #[derive(Deserialize)]
    struct Vector {
        sample: u32,
        cipher: String,
        key_hex: String,
        radix: u32,
        tweak_hex: String,
        plaintext: String,
        ciphertext: String,
    }

    fn unhex(s: &str) -> Vec<u8> {
        assert!(s.len() % 2 == 0);
        (0..s.len())
            .step_by(2)
            .map(|i| u8::from_str_radix(&s[i..i + 2], 16).unwrap())
            .collect()
    }

    fn to_numerals(s: &str) -> Vec<u8> {
        s.chars()
            .map(|c| NIST_ALPHABET.find(c).expect("char in NIST alphabet") as u8)
            .collect()
    }

    fn to_string(xs: &[u8]) -> String {
        xs.iter()
            .map(|&x| NIST_ALPHABET.as_bytes()[x as usize] as char)
            .collect()
    }

    /// Every official NIST FF1-AES sample, encrypt and decrypt directions.
    #[test]
    fn nist_ff1_sample_vectors() {
        let path = concat!(env!("CARGO_MANIFEST_DIR"), "/../testvectors/ff1_nist.json");
        let file: File =
            serde_json::from_str(&std::fs::read_to_string(path).expect("read ff1_nist.json"))
                .expect("parse ff1_nist.json");
        assert_eq!(file.vectors.len(), 9);

        for v in &file.vectors {
            let key = unhex(&v.key_hex);
            let tweak = unhex(&v.tweak_hex);
            let cipher = AesCipher::new(&key);
            let p = params(v.radix, &tweak, v.plaintext.chars().count());

            let pt = to_numerals(&v.plaintext);
            let ct = to_numerals(&v.ciphertext);

            let got_ct = encrypt(&cipher, &p, v.radix, &pt);
            assert_eq!(
                to_string(&got_ct),
                v.ciphertext,
                "sample {} ({}) encrypt",
                v.sample,
                v.cipher
            );

            let got_pt = decrypt(&cipher, &p, v.radix, &ct);
            assert_eq!(
                to_string(&got_pt),
                v.plaintext,
                "sample {} ({}) decrypt",
                v.sample,
                v.cipher
            );
        }
    }

    /// Encrypt/decrypt must be inverses for assorted lengths and radices.
    #[test]
    fn ff1_roundtrip() {
        let cipher = AesCipher::new(&[7u8; 16]);
        for &(radix, n) in &[(10u32, 6usize), (16, 16), (36, 19), (62, 12), (94, 4), (2, 20)] {
            let p = params(radix, b"dealcode/v1/test", n);
            for seed in 0u128..50 {
                let x: Vec<u8> = (0..n)
                    .map(|i| ((seed.wrapping_mul(2654435761) as usize + i * 7) % radix as usize) as u8)
                    .collect();
                let y = encrypt(&cipher, &p, radix, &x);
                assert_eq!(y.len(), n);
                assert_eq!(decrypt(&cipher, &p, radix, &y), x);
            }
        }
    }
}
